src/ClpModel.hpp

```diff
@@ -20,8 +21,13 @@
 #include "CoinHelperFunctions.hpp"
 #include "CoinTypes.h"
 #include "CoinFinite.hpp"
-#include "ClpParameters.hpp"
+#include "ClpModelParameters.hpp"
 #include "ClpObjective.hpp"
+
+#ifdef COINUTILS_HAS_GLPK
+#include "glpk.h"
+#endif
```

Comment:
Similar problem as with Osi: This means that if CoinUtils was build with Glpk, then both the build of Clp and every user who includes ClpModel.hpp in their code needs to make sure that glpk.h is found.
There is a reason why even the OsiGlpk header avoided including glpk.h.

Reply:
It does look to me now as though we don't need to include `glpk.h` here and I see the point that, in general, we should not need to include headers of packages that are only indirect dependencies. Nevertheless, I don't actually see how what you're suggesting can happen, so just for my better understanding, can you be more specific about the scenario in which this could happen? Here is the logic of why I don't see that it can (unless someone intentionally does something to violate the principle of how packaging is supposed to work).

- If Glpk is installed through the ThirdParty package, then it should actually guaranteed that `glpk.h` is in the include path, since `coinglpk` is a dependency of CoinUtils and thre an indirect dependency of Osi, so the include path will be added when parsing `coinglpk.pc` (which happens by pulling in CoinUtils, even if Osi itself does not pull in Glpk).
- On the other hand, if Glpk is linked by installing it separately and not using the ThirdParty package, then (I think) the include path should be added when parsing `coinutils.pc`. After all, we have to be able to include headers from CoinUtils that themselves include "glpk.h" when link to a CoinUtils that was built with Glpk support, even if Osi itself is not built with such support for some reason. 

The only weird case I can see is if CoinUtils is built on a machine that has Glpk and then packaged as a binary and installed on another machine without Glpk, it is possible that one might not include any CoinUtils headers that would cause a build failure, but then encounter a build failure here. But the whole pkgconfig scheme and the way package managers work across the board is meant to prevent this case. Do you see what I mean?

Reply:
We actually do need to include it here, since there is a `readGMPL` function in the ClpModel class that passes the GMPL data structures. In other words, Glpk is a direct dependency, not an indirect one. So we should technically be checking for Glpk in the configure of Clp and then I guess we need to be sure here that both COINUTILS_HAS_GLPK and CLP_HAS_GLPK are defined, since we need both the direct dependency and the indirect one. 

We can avoid all of this in Osi because there, we only call the CoinMpsIO::readGMPL, and we leave off the two optional arguments that are for passing the GMPL data structures in and out (you don't need these if you are happy with what is passed out through the MpsIO object). So this means Glpk is only an indirect dependency in Osi.

Clp currently uses the GMPL data structures directly, but it's possible it could be avoided by adding whatever information Clp needs that is not stored in the MpsIO object to the MpsIO object in CoinUtils, so that we can use the same trick as in Osi to make Glpk only an indirect dependency. That actually sounds like the right thing to do, but I'm not sure how much work it would be, so I'm going to leave this as is for now and we can address it separately if we want.

Reply:
I'm never sure in which case pkg-config will decide to add the CFLAGS also from dependencies. `coinglpk` is under `Requires.private` for CoinUtils, which I would think could mean that it doesn't need to make the CFLAGS of coinglpk available to CoinUtils users. But, ok, maybe that is wrong. The other scenario where the glpk header is no longer available when someone builds against a redistributed CoinUtils lib is possible, yes.

We took effort in the past to not include glpk.h in API header files. It just seems so unnecesary to undo all this. The `readGMPL` function does not get GMPL data structures as arguments but **pointers to**  GMLP data structures. That is, to compile the header, the structure itself does not need to be known.
OsiGlpk handles this by having a
```c
typedef struct glp_prob glp_prob;
```
in its header file and I don't understand why this isn't done for all the GMPL processing, too.

Reply:
> We actually do need to include it here, since there is a `readGMPL` function in the ClpModel class that passes the GMPL data structures. In other words, Glpk is a direct dependency, not an indirect one. So we should technically be checking for Glpk in the configure of Clp and then I guess we need to be sure here that both COINUTILS_HAS_GLPK and CLP_HAS_GLPK are defined, since we need both the direct dependency and the indirect one.
> 
> We can avoid all of this in Osi because there, we only call the CoinMpsIO::readGMPL, and we leave off the two optional arguments that are for passing the GMPL data structures in and out (you don't need these if you are happy with what is passed out through the MpsIO object). So this means Glpk is only an indirect dependency in Osi.
> 
> Clp currently uses the GMPL data structures directly, but it's possible it could be avoided by adding whatever information Clp needs that is not stored in the MpsIO object to the MpsIO object in CoinUtils, so that we can use the same trick as in Osi to make Glpk only an indirect dependency. That actually sounds like the right thing to do, but I'm not sure how much work it would be, so I'm going to leave this as is for now and we can address it separately if we want.

Encapsulating all GMPL use in CoinUtils is hopefully on someones todo list. Similar with the AMPL stuff. Having this spread around different projects makes things difficult to maintain. As you say, you will have to repeat checks for GLPK or ASL, deal with the case that you have a  CoinUtils build without GLPK but a Clp with Glpk, etc.

Reply:
The dependence on GLPK looks non-trivial and not that easy to encapsulate back in Coinutils, so for now, I just added a check for Glpk in Clp. To compile readGmpl, it is now required that both Clp and CoinUtils are built with Glpk support. Fixed now in a93d1d8b.